### PR TITLE
Add some timing tests for trackedFunction

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Investigation of https://github.com/NullVoxPopuli/ember-resources/issues/1010

### CODE REVIEW

File: ember-resources/src/util/function.ts
Comment (NullVoxPopuli) [inline ember-resources/src/util/function.ts:278]: this is the fix

**Comment by NullVoxPopuli** [comment]
Blocked on https://github.com/actions/setup-node/pull/865

File: ember-resources/src/util/function.ts
Comment (Techn1x) [reply ember-resources/src/util/function.ts:278]: This makes a lot of sense. And these code comments are great

File: test-app/tests/utils/function/rendering-test.gts
Comment (Techn1x) [inline test-app/tests/utils/function/rendering-test.gts:10]: Looks like this cell import is unused, probably leftover from investigation

File: test-app/tests/utils/function/timing-test.gts
Comment (Techn1x) [inline test-app/tests/utils/function/timing-test.gts:22]: :smile:

**Review by Techn1x — APPROVED** [review]
Thanks a bunch. Tests are great too!

**Comment by Techn1x** [comment]
I'll find some time today to pnpm patch the fix in my app to test it out, but I'm pretty confident you nailed it

File: test-app/tests/utils/function/rendering-test.gts
Comment (NullVoxPopuli) [reply test-app/tests/utils/function/rendering-test.gts:10]: thanks, yeah, I need to fix linting on this repo :upside_down_face: 
there is a bug with prettier rn though, and I'm waiting for them to do another release

**Comment by Techn1x** [comment]
Seems that the fix line causes a tracking error 🤔 

Screenshot 2023-10-05 at 10 11 02 am

I can see the patch I added, so I know it's being applied

Screenshot 2023-10-05 at 10 13 11 am

**Comment by NullVoxPopuli** [comment]
ooo this is good to catch before merging this -- what's the code that caused that?

**Comment by Techn1x** [comment]
The error points to line 188 in that minified source which looks like it's the `if (this.data) this.data = null` fix line that produces the error.

It seems to be on initial run. Looks like retry() function gets called twice for some reason

**Comment by Techn1x** [comment]
> what's the code that caused that?

```ts
class TimeOnTaskUsage extends Component<Signature> {
  @service studentEvents!: StudentEventsService

  @service store!: Store

  fetchUsage = trackedFunction(this, () => {
    return this.studentEvents.timeOnTaskUsage( // returns a promise
      'writing_legends',
      this.args.statsScope,
      this.args.period,
    )
  })

  <template>
    {{#if this.fetchUsage.isPending}}
      loading!
    {{else if this.fetchUsage.isError}}
      error!
    {{else if this.fetchUsage.value}}
      {{log this.fetchUsage.value}}
    {{/if}}
  </template>
```

**Comment by NullVoxPopuli** [comment]
how are you causing the change to happen? statsScope changing? both args? something else?

thanks!

**Comment by Techn1x** [comment]
`@tracked period = 'named-period:this-year'` is a controller tracked query param that is passed down to this component
`statsScope` is a getter derived from a value provided by the route model
```ts
  get statsScope(): StatsScope {
    if (this.args.model.uiScope.scope === 'school') {
      return { scope: 'school', ids: [this.args.model.uiScope.id] }
    } else {
      ...
    }
  }
```

But I think that's a red herring - I can still repro the error even when the tracked values are removed from the function, since the error occurs on first run.
```ts
  fetchUsage = trackedFunction(this, () => {
    return this.studentEvents.timeOnTaskUsage( // returns a promise
      'writing_legends',
      { scope: 'school', ids: ['1'] }, // this.args.statsScope,
      'named-period:this-year', // this.args.period,
    )
  })
```

The fact that your tests seem to work tells me that it's probably something unique to my setup, even though this is a fairly simple repro in my app :/

I will try to get this reproduced in some kind of playground

**Comment by NullVoxPopuli** [comment]
ya know, I get the error now, too.
This is good. It's all over my tests.

So I think I fell for the `injected` peerDependency problem again :weary: 
it seems my last push caused this, and the push before was fine. I've reverted that change now.

apologies for the run around.
it would have popped up when the setup-node released a new version -- but oofta.

**Comment by Techn1x** [comment]
> apologies for the run around.

no trouble at all, I'm just glad to help! You do so much cool work!

I ended up coding up a potential repro in limber, though - but since it's using this addon before this PR fix, it doesn't show the error, but I was able to manually repro the bad behaviour by debugging in the console and running `if (this.data) this.data = null` at the correct place in the minified code

(and sure enough, if I instead run `this.data = null` at the same place without the data read, the bad behaviour goes goes away)

not sure if this is the correct way to share, but it seems to share
https://limber.glimdown.com/edit?c=IYFxFMFsAcIEwAQCdzSQewegZgkTgBjAawEsA7AcwVIGdaBXcBACzGloC4B6by0kCwYAjAHSF0kbgDkGAGzkA1dAA8ACumjzS3KMPBIAtClroGSQuFrctC7gEYADPfsAoVwAMvlAFa0EcqQAbuCupDDoSCAIAMKS0Ojk4OTR2BiQCADkAAKUgZCQBtwSEUkpmQDcrrw0EVEIAN54BCTgiAC%2BCGmSWbn5hUjc%2BERkVJXu4Qn1TcOtcABiDOSEIKSJCJ3dGZl6BsZWZhZW3AyrctzYSytr5OOu4CpT0XDg2MDy0YRywPQIACJQTAPCDkOD%2BOKlZLRBquBBdcAgQgsACqtGAlGYAF5miM2otlqtEgAKQR0AA0CB%2BAE9lggiQBKBCYgB8jVhcOQCPM5EpAHdgAIEEleQg1Ok6OAiUSTIyWQhaAiACrhcBmEBS2WskwkpBMekUpyORz0%2Bns9r0qrsgA8EBg3wgzPZcIaDQAxKRcKTaKJsAikaj0eBRHQ1Mk4BRKO12k6OQAZADyAEE-gBJaQAcRjLvAcgVNE9LDoPr9KLRGODtAAokgMEgozG4ZWAEpN%2BNNrMNHN5j14Qve32I0uB0RBYByJj1jkcgDKyJiMUr0%2BnHe4HsnCCtQyg0Ht4Ed0a8HiAA&format=glimdown

**Comment by Techn1x** [comment]
confirmed the updated fix works in my app too 🚀

**Pull request merged**
